### ISSUE TITLE
Cannot read property 'command' of undefined

### ISSUE THREAD

**Issue by astevelinocermati** [mongodb-js/mongodb-core#216]
Hi, i'm using mongoose to build my app, it ran smoothly until we upgrade to mongodb 3.4. But there's one query which throw error at some point. This error seems not appear every time, most of the queries ran successfully. Something wrong with wire protocol handler? Thanks for the support.
```
try {
      self.query = self.server.wireProtocolHandler.command(self.bson, self.ns, self.cmd, self.cursorState, self.topology, self.options);
    } catch(err) {
      return callback(err);
    }
```
```
TypeError: Cannot read property 'command' of undefined
    at nextFunction (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:568:51)
    at Cursor.next [as _next] (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:692:3)
    at fetchDocs (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/cursor.js:856:10)
    at toArray (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/cursor.js:883:3)
    at Cursor.toArray (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/cursor.js:836:44)
    at /home/appRunner/workspace/node_modules/mongoose/node_modules/mquery/lib/collection/node.js:29:12
    at handleCallback (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/utils.js:120:56)
    at Collection.find (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/collection.js:385:44)
    at NativeCollection.(anonymous function) [as find] (/home/appRunner/workspace/node_modules/mongoose/lib/drivers/node-mongodb-native/collection.js:126:28)
    at NativeCollection.Collection.doQueue (/home/appRunner/workspace/node_modules/mongoose/lib/collection.js:123:30)
    at NativeCollection.Collection.onOpen (/home/appRunner/workspace/node_modules/mongoose/lib/collection.js:83:8)
    at callback (/home/appRunner/workspace/node_modules/mongoose/lib/drivers/node-mongodb-native/collection.js:85:43)
    at NativeCollection.onOpen (/home/appRunner/workspace/node_modules/mongoose/lib/drivers/node-mongodb-native/collection.js:43:5)
    at open (/home/appRunner/workspace/node_modules/mongoose/lib/connection.js:678:30)
    at /home/appRunner/workspace/node_modules/mongoose/lib/connection.js:689:7
    at /home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/authenticate.js:96:5
    at handleCallback (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/utils.js:120:56)
    at _callback (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/authenticate.js:23:5)
    at /home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/lib/authenticate.js:66:7
    at /home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/replset.js:1227:9
    at /home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/pool.js:758:7
    at /home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/pool.js:736:11
    at finish (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/auth/scram.js:160:16)
    at handleEnd (/home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/auth/scram.js:173:7)
    at /home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/auth/scram.js:270:13
    at /home/appRunner/workspace/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/pool.js:469:18
    at nextTickCallbackWith0Args (node.js:436:9)
    at process._tickCallback (node.js:365:13)
```

**Comment by mbroadst** [comment]
@astevelinocermati the `wireProtocolHandler` is only assigned in two locations: 
  * https://github.com/mongodb-js/mongodb-core/blob/2.0/lib/topologies/server.js#L128
  * https://github.com/mongodb-js/mongodb-core/blob/2.0/lib/topologies/server.js#L276

And otherwise is always assigned a value, so there must be some sort of race condition during (my guess) the second assignment linked to above.  It would be incredibly helpful if you could somehow provide code that reproduced this error and/or could provide more details about your environment when you are experiencing this error.  Thanks!

**Comment by astevelinocermati** [comment]
Thanks for the reply. That's one possible condition. Actually when we try to start keystone, this error also occurs sometimes. We still figuring out what race condition possibilities from our code. So our biggest suspect from now is race condition when we try to start keystone.

I will update this thread once we have found the problem, so everyone can learn from our mistake. Once again thanks for the support.

**Comment by astevelinocermati** [comment]
@mbroadst 
After several trials and console.log, I found something interesting. The pattern to trigger this error is when after mongodb **authenticate**, `self.ismaster` is primary replica set. I tried to console.log `self.ismater` before https://github.com/mongodb-js/mongodb-core/blob/2.0/lib/topologies/server.js#L276. If after **authenticate** `self.ismaster` filled with secondary replica set, then no error was found.

That explains a lot no error was found during development since we didn't use replica set (only production use replica). But what i want to clarify is, does it have relation between primary and secondary replica set with this `wireProtocolHandler`.

One more thing. if after **authenticate**, `self.ismaster` is primary, we found that https://github.com/mongodb-js/mongodb-core/blob/2.0/lib/topologies/server.js#L128 is not being called at all.

***authenticate**, refer to this: https://github.com/mongodb/node-mongodb-native/blob/2.2/lib/db.js#L1485

**Comment by mbroadst** [comment]
@astevelinocermati very interesting! Sounds like we have enough to make a unit test to express this bug.  We're still in the process of implementing 3.6 features, but this is on the top of the list otherwise - thanks for the triaging

**Comment by astevelinocermati** [comment]
@mbroadst it looks like our team has found something. With mongoose, we can set db readPreference (refer to this http://mongodb.github.io/node-mongodb-native/driver-articles/mongoclient.html#read-preference) and we found that our code set the readPreference with `undefined`. We changed it to `'secondary'` and magically, the race condition is over. I still don't know why, but it works. Thanks.
